pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -1060,7 +1060,29 @@ func isKernelPid(pid int) bool {
 	return err != nil && os.IsNotExist(err)
 }
 
+// GetCapacity returns node capacity data for "cpu", "memory", "ephemeral-storage", and "huge-pages*"
+// At present this method is only invoked when introspecting ephemeral storage
 func (cm *containerManagerImpl) GetCapacity() v1.ResourceList {
+	// We store allocatable ephemeral-storage in the capacity property once we Start() the container manager
+	if _, ok := cm.capacity[v1.ResourceEphemeralStorage]; !ok {
+		// If we haven't yet stored the capacity for ephemeral-storage, we can try to fetch it directly from cAdvisor,
+		if cm.cadvisorInterface != nil {
+			rootfs, err := cm.cadvisorInterface.RootFsInfo()
+			if err != nil {
+				klog.ErrorS(err, "Unable to get rootfs data from cAdvisor interface")
+				// If the rootfsinfo retrieval from cAdvisor fails for any reason, fallback to returning the capacity property with no ephemeral storage data
+				return cm.capacity
```

Comment:
what is the possible reason for it to fail? 
Is it better to retry here a few times instead of returning empty data?

Reply:
Failure conditions are the low likelihood that the root volume on the OS does not return a proper result when introspecting for available/allocatable storage (very rare I/O edge case and/or catastrophic OS failure) or a result of cAdvisor runtime being stuck in some way (not reproduced in practice.

tl;dr this should not happen, and in my real-world tests I have never seen it happen

Reply:
@cynepco3hahue @ehashman thoughts? Do we think a retry window of ~1-2 seconds is appropriate? (As well as the add'l go routine overhead that a async retry loop will require?)

Thx for the review @jingxu97!

Reply:
Per https://github.com/kubernetes/kubernetes/issues/99305#issuecomment-794926298 I think worrying that cadvisor won't be ready yet is a valid concern... cc @pacoxu

Reply:
According to this PR (disclaimer: from 4 years ago!) it is not:

https://github.com/kubernetes/kubernetes/pull/48567

But we're also setting storage capacity differently right now than in that above PR. (Although, still not setting it in the same flow as CPU/memory).

It does not seem that cadvisor would return a "wrong" value, rather that it would fail to get storage data from the "/" volume and return an error.

Reply:
Thanks for normalizing everyone's work here @ehashman :)

I don't see how this PR, as-is, could introduce further race conditions. It is simply adding a novel "best effort" implementation of gathering storage data, doing it the exact same way we do it now during containerManager Start(). Without any retries, this won't *guarantee* that we get valid data, but in my tests it definitely reduces the occurence of "get node info" code flows that result in a node representation with zero allocatable ephemeral-storage data.

Reply:
> I don't see how this PR, as-is, could introduce further race conditions

The race would be:

- we see the storage capacity isn't set
- underneath us, storage capacity gets set
- but we are not setting it atomically, so we check cadvisor here, and potentially overwrite it
- and, if we fail to check cadvisor here, but this was already set elsewhere, we end up overwriting it with a 0

Reply:
Thanks for the detail! If I understand this correctly I believe this is being considered already:

https://github.com/kubernetes/kubernetes/blob/bcfe175c3dd756eba96b05457445a279c5f8a65a/pkg/kubelet/cm/container_manager_linux.go#L1076

No change is being made to the container manager's repesentation of storage capacity. We are doing a "best effort" poll of cAdvisor during the node "get storage capacity" flow *if the container manager does not (yet) have storage capacity data". We are doing that in a "best effort" way because we don't want to slow down kubelet bootstrapping operations (for example, by retrying node data gathering until storage is available). That "best effort" tactic is due to review comments here. Happy to consider a more definitive solution if slowing down kubelet bootstrapping is an option. I think this PR as-is is a reasonable trade off given the undesirable current behaviors (and the complexity of the kubelet bolognese around these parts 😄).

Reply:
Checking this over again, it's probably fine because we're returning the pointer rather than a value on error.